src/plugins/plugin.gridlines.js

```diff
@@ -0,0 +1,166 @@
+			horizontal: {},
+			vertical: {}
+		};
+
+		// Marks scale border positions to prevent overlapping of gridLines and scale borders
+		helpers.each(chart.scales, function(scale) {
+			var scaleOptions = scale.options;
+
+			// gridLines.drawBorder is deprecated
+			if (scaleOptions.gridLines.drawBorder && scaleOptions.borderColor !== null && scaleOptions.borderWidth !== 0 && scaleOptions.borderWidth !== null) {
+				var glHashByOrientantion = gridLinesHash[!scale.isHorizontal() ? 'horizontal' : 'vertical'];
+				var borderPosition;
+
+				if (scale.isHorizontal()) {
+					borderPosition = scale.position === 'top' ? scale.bottom : scale.top;
+				} else {
+					borderPosition = scale.position === 'left' ? scale.right : scale.left;
+				}
+
+				glHashByOrientantion[Math.round(borderPosition)] = true;
+			}
+		});
+
+		// Collects gridLines
+		helpers.each(chart.scales, function(scale) {
+			var scaleOptions = scale.options;
+
+			if (scaleOptions.display && scaleOptions.gridLines.display && scaleOptions.gridLines.drawOnChartArea) {
+				var glHashByOrientantion = gridLinesHash[scale.isHorizontal() ? 'horizontal' : 'vertical'];
+				var position;
+
+				for (var tickIndex = 0; tickIndex < scale.ticks.length; tickIndex++) {
+					var tick = scale.ticks[tickIndex];
+
+					if (tick === null || tick === undefined) {
+						continue;
+					}
+
+					position = scale.getPixelForTick(tickIndex);
+
+					if (glHashByOrientantion[position] === undefined) {
+						glHashByOrientantion[position] = true;
+						lines.push(getGridLine(chart, scale, position, tickIndex));
+					}
+				}
+
+				// When offsetGridLines is enabled, there should be one more gridLine than there
+				// are ticks in scale.ticks array, thefore the missing gridLine has to be manually added
+				if (scaleOptions.gridLines.offsetGridLines) {
+					position = Math.round(!scale.isHorizontal() ? scale.bottom : scale.right);
+
+					if (glHashByOrientantion[position] === undefined) {
+						glHashByOrientantion[position] = true;
+						lines.push(getGridLine(chart, scale, position, undefined));
+					}
+				}
+			}
+		});
+	}
+
+	function drawGridLines(chart) {
+		var context = chart.ctx;
+
+		for (var i = 0; i < lines.length; i++) {
+			var line = lines[i];
+
+			context.lineWidth = line.width;
+			context.strokeStyle = line.color;
+			if (context.setLineDash) {
+				context.setLineDash(line.borderDash);
+				context.lineDashOffset = line.borderDashOffset;
+			}
+
+			var aliasPixel = helpers.aliasPixel(context.lineWidth);
+			var x1 = line.x1;
+			var x2 = line.x2;
+			var y1 = line.y1;
+			var y2 = line.y2;
+
+			if (y1 === y2) {
+				y1 += aliasPixel;
+				y2 += aliasPixel;
+			} else {
+				x1 += aliasPixel;
+				x2 += aliasPixel;
+			}
+
+			context.beginPath();
+
+			context.moveTo(x1, y1);
+			context.lineTo(x2, y2);
+
+			context.stroke();
+			context.restore();
+		}
+	}
+
+	return {
+		id: 'gridlines',
+
+		afterUpdate: function(chart) {
+			collectVisibleGridLines(chart);
```

Comment:
You need to store the `lines` per chart and a good way to do that is to add an plugin specific private model under `chart` (see [this example](/chartjs/chartjs-plugin-deferred/blob/master/src/plugin.js#L132)).

```javascript
var MODEL_KEY = '$gridlines';

module.exports = function(Chart) {
    // ...

    return {
        afterUpdate: function(chart) {
            chart[MODEL_KEY] = {
                lines: collectVisibleGridLines(chart);
            };
        }

        beforeDraw: function(chart) {
            var model = chart[MODEL_KEY];
            if (model) {
                drawGridLines(chart.ctx, model.lines);
            }
        }
    };
}
```